platform/android/export/export.cpp

```diff
@@ -2520,34 +2587,12 @@ class EditorExportPlatformAndroid : public EditorExportPlatform {
 
 		Vector<String> enabled_abis = get_enabled_abis(p_preset);
 
-		String project_icon_path = ProjectSettings::get_singleton()->get("application/config/icon");
-
 		// Prepare images to be resized for the icons. If some image ends up being uninitialized, the default image from the export template will be used.
 		Ref<Image> launcher_icon_image;
 		Ref<Image> launcher_adaptive_icon_foreground_image;
 		Ref<Image> launcher_adaptive_icon_background_image;
 
-		launcher_icon_image.instance();
-		launcher_adaptive_icon_foreground_image.instance();
-		launcher_adaptive_icon_background_image.instance();
-
-		// Regular icon: user selection -> project icon -> default.
-		String path = static_cast<String>(p_preset->get(launcher_icon_option)).strip_edges();
-		if (path.empty() || ImageLoader::load_image(path, launcher_icon_image) != OK) {
-			ImageLoader::load_image(project_icon_path, launcher_icon_image);
-		}
-
-		// Adaptive foreground: user selection -> regular icon (user selection -> project icon -> default).
-		path = static_cast<String>(p_preset->get(launcher_adaptive_icon_foreground_option)).strip_edges();
-		if (path.empty() || ImageLoader::load_image(path, launcher_adaptive_icon_foreground_image) != OK) {
-			launcher_adaptive_icon_foreground_image = launcher_icon_image;
-		}
-
-		// Adaptive background: user selection -> default.
-		path = static_cast<String>(p_preset->get(launcher_adaptive_icon_background_option)).strip_edges();
-		if (!path.empty()) {
-			ImageLoader::load_image(path, launcher_adaptive_icon_background_image);
-		}
+		load_icon_refs(p_preset, launcher_icon_image, launcher_adaptive_icon_foreground_image, launcher_adaptive_icon_background_image);
```

Comment:
Seems like this should be behind a `if(!use_custom_build)` flag, otherwise when `use_custom_build` is `true` the images get loaded twice.

An alternative would be to load the images earlier in the method and to update the method signature for `_copy_icons_to_gradle_project` so you can pass the image parameters.

Reply:
Should I move this down to line 2630 and put it behind the same `if (!use_custom_build)` flag? Or is it better to refactor the `_copy_icons_to_gradle_project` method? Personally, I prefer the first option.

Reply:
@amanj120 refactoring `_copy_icons_to_gradle_project` is preferable, otherwise whenever this method is updated and the image variables referenced, it's easy to overlook the fact that the variables are only loaded when `use_custom_build` is false which could lead to subtle bugs.